### PR TITLE
Add model from RCM

### PR DESCRIPTION
**Pull request by erikwijmans**
Adds the model from RCM paper

Adds caching of visual features for faster training

### CODE REVIEW

File: habitat_baselines/il/vln/dagger.py [hunk 1]
```diff
@@ -343,12 +370,13 @@ def _update_dataset(self, data_it):
                             np.array([step[1] for step in ep], dtype=np.int64),
                             np.array([step[2] for step in ep], dtype=np.int64),
                         ]
-                        txn.put(
-                            str(start_id + collected_eps).encode(),
-                            msgpack_numpy.packb(
-                                transposed_ep, use_bin_type=True
-                            ),
-                        )
+                        with lmdb_env.begin(write=True) as txn:
```
Comment (erikwijmans) [inline habitat_baselines/il/vln/dagger.py]: This was moved here as that'll cause it to write each thing immediately and reduces RAM usage (I was scared of ooming on 60GB).

File: habitat_baselines/models/resnet_encoders.py [hunk 2]
```diff
@@ -226,21 +283,47 @@ def forward(self, observations):
...
+                )
+                .view(1, -1, h, w)
+                .expand(b, self.spatial_embeddings.embedding_dim, h, w)
+            )
+
+            return torch.cat([resnet_output, spatial_features], dim=1)
+        else:
+            return self.activation(
+                self.fc(resnet_output)
+            )  # [BATCH x OUTPUT_DIM]
```
Comment (jacobkrantz) [inline habitat_baselines/models/resnet_encoders.py]: This linear layer expects resnet_output to be `[BATCH x linear_layer_input_size]` but when I ran it with VLNBaselinePolicy, resnet_output was sometimes `[BATCH x linear_layer_input_size x 1 x 1]` and it errored out. We could maybe squeeze and check if batch if batch size is 1:

            resnet_output = resnet_output.squeeze()
            if len(resnet_output.size()) == 1:
                resnet_output = resnet_output.unsqueeze(0)

Both VLNBaselinePolicy and VLNRCMPolicy ran on my machine once this was in place.

File: habitat_baselines/il/vln/dagger.py [hunk 3]
```diff
@@ -409,9 +449,14 @@ def _update_dataset(self, data_it):
                 )
                 batch = batch_obs(observations, self.device)
 
+            txn.commit()
+
         self.envs.close()
         self.envs = None
 
+        rgb_hook.remove()
+        depth_hook.remove()
```
Comment (jacobkrantz) [inline habitat_baselines/il/vln/dagger.py]: depth_hook may be None if we don't use VlnResnetDepthEncoder

**Review by jacobkrantz — CHANGES_REQUESTED** [review]
Small change for compatibility and its running. This is great!

File: habitat_baselines/models/resnet_encoders.py (hunk 2 above)
Comment (erikwijmans) [reply habitat_baselines/models/resnet_encoders.py]: Ah, so it happens when caching features.  The cached features don't get the squeeze.  I removed the flatten from the hook and added a call to it before the FC layer.  That felt cleaner than squeezing the features to cache for some reason :)

**Pull request merged**